packages/search-input/src/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,377 @@
+
+import {
+  baseInputStyle,
+  clearButtonSizeStyle,
+  formStyle,
+  inputSizeStyles,
+  inputThemeStyle,
+  inputWrapperDisabledStyle,
+  inputWrapperFocusStyles,
+  inputWrapperInteractiveStyles,
+  inputWrapperSizeStyle,
+  inputWrapperStyle,
+  inputWrapperThemeStyle,
+  searchIconDisabledStyle,
+  searchIconSizeStyle,
+  searchIconStyle,
+  searchIconThemeStyle,
+} from './SearchInput.styles';
+import { SearchInputProps, SizeVariant, State } from './SearchInput.types';
+
+/**
+ * # SearchInput
+ *
+ * SearchInput component
+ *
+ * @param props.state The current state of the SearchInput. This can be none, or loading.
+ * @param props.darkMode determines whether or not the component appears in dark theme.
+ * @param props.sizeVariant determines the size of the text and the height of the input.
+ * @param props.baseFontSize determines the base font size if sizeVariant is set to default.
+ */
+
+export const SearchInput = React.forwardRef<HTMLInputElement, SearchInputProps>(
+  function SearchInput(
+    {
+      placeholder = 'Search',
+      className,
+      darkMode: darkModeProp,
+      sizeVariant = SizeVariant.Default,
+      disabled,
+      children,
+      state = State.Unset,
+      value: valueProp,
+      onChange: onChangeProp,
+      onSubmit: onSubmitProp,
+      ...rest
+    }: SearchInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) {
+    const { theme, darkMode } = useDarkMode(darkModeProp);
+    const [isOpen, setOpen] = useState(false);
+    // The index of the currently highlighted result option
+    const [highlightIndex, setHighlightIndex] = useState<number>(0);
+
+    const closeMenu = () => setOpen(false);
+    const openMenu = () => setOpen(true);
+
+    const formRef = useRef<HTMLFormElement>(null);
+    const searchBoxRef = useRef<HTMLDivElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const inputRef = useForwardedRef(forwardRef, null);
+    const clearButtonRef = useRef<HTMLButtonElement>(null);
+    const resultRefs = useDynamicRefs<HTMLElement>({ prefix: 'result' });
+    const withTypeAhead = !isUndefined(children);
+    const [focusedElement, trackFocusedElement] = useState<Element>();
+
+    const { value, handleChange } = useControlledValue(valueProp, onChangeProp);
+
+    /**
+     * Helper function that both counts the number of `SearchResult` descendants
+     * and adds the appropriate props to said children
+     */
+    const processChildren = () => {
+      let resultsCount = 0;
+
+      const processChild = (
+        child: React.ReactNode,
+      ): JSX.Element | undefined => {
+        if (isComponentType(child, 'SearchResult')) {
+          resultsCount += 1;
+          const index = resultsCount - 1;
+          return React.cloneElement(child, {
+            ...child.props,
+            id: `result-${index}`,
+            ref: child.props.ref ?? resultRefs?.(`${index}`),
+            focused: index === highlightIndex,
```

Comment:
should we be passing an onClick here, such that when an item is clicked the menu closes?

Reply:
Any `onClick` handlers the user has added would be included in `child.props`. Also, since in most cases the search results will navigate to a new page, it's not absolutely necessary to force the menu to close